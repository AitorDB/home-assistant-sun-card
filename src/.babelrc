{
  "presets": [
    "@babel/env",
    "@babel/typescript"
  ],
  "plugins": [
    ["@babel/plugin-proposal-decorators", { "decoratorsBeforeExport": true }],
    "@babel/proposal-class-properties"
  ]
}
