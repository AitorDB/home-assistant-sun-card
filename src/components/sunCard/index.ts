export * from './SunCard'
export * from './SunCardContent'
export * from './SunCardFooter'
export * from './SunCardGraph'
export * from './SunCardHeader'
