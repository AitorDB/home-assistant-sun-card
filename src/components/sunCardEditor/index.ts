export * from './SunCardEditor'
export * from './SunCardEditorContent'
