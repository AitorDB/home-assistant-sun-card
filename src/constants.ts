import da from './assets/localization/languages/da.json'
import de from './assets/localization/languages/de.json'
import en from './assets/localization/languages/en.json'
import es from './assets/localization/languages/es.json'
import et from './assets/localization/languages/et.json'
import fi from './assets/localization/languages/fi.json'
import fr from './assets/localization/languages/fr.json'
import hu from './assets/localization/languages/hu.json'
import it from './assets/localization/languages/it.json'
import lt from './assets/localization/languages/lt.json'
import nl from './assets/localization/languages/nl.json'
import pl from './assets/localization/languages/pl.json'
import ptBR from './assets/localization/languages/pt-BR.json'
import ru from './assets/localization/languages/ru.json'
import sl from './assets/localization/languages/sl.json'
import sv from './assets/localization/languages/sv.json'
import { TSunCardConfig, TSunCardTexts } from './types'

export class Constants {
  static readonly DEFAULT_CONFIG: TSunCardConfig = {
    darkMode: true,
    language: 'en',
    showAzimuth: false,
    showElevation: false,
    timeFormat: '24h'
  }

  static readonly EVENT_X_POSITIONS = {
    dayStart: 5,
    sunrise: 101,
    sunset: 449,
    dayEnd: 545
  }

  static readonly HORIZON_Y = 108
  static readonly LOCALIZATION_LANGUAGES: Record<string, TSunCardTexts> = {
    da, de, en, es, et, fi, fr, hu, it, lt, nl, pl, 'pt-BR': ptBR, ru, sl, sv
  }
  static readonly SUN_RADIUS = 17
}
