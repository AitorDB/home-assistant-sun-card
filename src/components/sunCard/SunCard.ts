import { HomeAssistant } from 'custom-card-helpers'
import { CSSResult, customElement, LitElement, state, TemplateResult } from 'lit-element'

import cardStyles from '../../cardStyles'
import { Constants } from '../../constants'
import { ESunCardErrors, ISunCardConfig, TSunCardData, TSunCardTimes, TSunInfo } from '../../types'
import { HelperFunctions } from '../../utils/HelperFunctions'
import { I18N } from '../../utils/I18N'
import { SunCardEditor } from '../sunCardEditor'
import { SunErrorContent } from '../SunErrorContent'
import { SunCardContent } from './SunCardContent'

@customElement('sun-card')
export class SunCard extends LitElement {
  static readonly cardType = 'sun-card'
  static readonly cardName = 'Sun Card'
  static readonly cardDescription = 'Custom card that display a graph to track the sun position and related events'

  @state()
  private config: ISunCardConfig = { type: SunCard.cardType }

  @state()
  private data!: TSunCardData

  private hasRendered = false
  private lastHass!: HomeAssistant

  static get styles (): CSSResult {
    return cardStyles
  }

  set hass (hass: HomeAssistant) {
    this.lastHass = hass

    if (!this.hasRendered) {
      this.populateConfigFromHass()
      return
    }

    this.processLastHass()
  }

  static getConfigElement (): HTMLElement {
    return document.createElement(SunCardEditor.cardType)
  }

  // called by HASS whenever config changes
  public setConfig (config: ISunCardConfig): void {
    const newConfig = { ...this.config }
    newConfig.title = config.title
    newConfig.darkMode = config.darkMode
    newConfig.language = config.language
    newConfig.use12hourClock = config.use12hourClock
    newConfig.component = config.component ?? Constants.DEFAULT_CONFIG.component

    if (newConfig.language && !HelperFunctions.isValidLanguage(newConfig.language)) {
      throw Error(`${config.language} is not a supported language. Supported languages: ${Object.keys(Constants.LOCALIZATION_LANGUAGES)}`)
    }

    const defaultFields = Constants.DEFAULT_CONFIG.fields!

    newConfig.fields = {
      sunrise: config.fields?.sunrise ?? defaultFields.sunrise,
      sunset: config.fields?.sunset ?? defaultFields.sunset,

      dawn: config.fields?.dawn ?? defaultFields.dawn,
      noon: config.fields?.noon ?? defaultFields.noon,
      dusk: config.fields?.dusk ?? defaultFields.dusk,

      azimuth: config.fields?.azimuth ?? defaultFields.azimuth,
      elevation: config.fields?.elevation ?? defaultFields.elevation
    }

    this.config = newConfig
    if (this.lastHass) {
      this.populateConfigFromHass()
    }
  }

  public render (): TemplateResult {
    if (this.data?.error) {
      return new SunErrorContent(this.config, this.data.error).render()
    }

    // TODO: Move
    // init i18n component (assume set config has run at least once)
    this.config.i18n = new I18N(this.config.language!, this.config.use12hourClock)

    // render components
    return new SunCardContent(this.config, this.data).render()
  }

  protected updated (changedProperties: Map<string | number | symbol, unknown>): void {
    super.updated(changedProperties)

    if (!this.hasRendered) {
      this.hasRendered = true
      this.processLastHass()
    }
  }

  private populateConfigFromHass () {
    // respect setting in hass
    // NOTE: custom-card-helpers types are not up to date with home assistant
    // NOTE: Old releases from Home Assistant doesn't provide the locale property
    this.config.darkMode = this.config.darkMode ?? (this.lastHass.themes as unknown as { darkMode: boolean })?.darkMode
    this.config.language = this.config.language ?? (this.lastHass as unknown as { locale?: { language: string }}).locale?.language ?? this.lastHass.language
  }

  private processLastHass () {
    if (!this.lastHass) {
      return
    }

    this.populateConfigFromHass()

    const sunComponent = this.config.component!

    if (this.lastHass.states[sunComponent]) {
      const sunAttrs = this.lastHass.states[sunComponent].attributes
      const times = this.readTimes(sunAttrs)

      const sunInfo = this.calculateSunInfo(times.sunrise, times.sunset)

      this.data = {
        azimuth: sunAttrs.azimuth,
        elevation: sunAttrs.elevation,
        sunInfo,
        times
      }

    } else {
      this.data = {
        azimuth: 0,
        elevation: 0,
        sunInfo: Constants.DEFAULT_SUN_INFO,
        times: Constants.DEFAULT_TIMES,
        error: ESunCardErrors.SunIntegrationNotFound
      }
    }
  }

  /* For the math to work in #calculateSunInfo(sunrise, sunset), we need the
   * date part of the given 'date-time' to be equal. This will not be the
   * case whenever we pass one of the 'times', ie: when we pass dawn, hass
   * will update that variable with tomorrows dawn.
   *
   * This function safe-guards that through standardizing the 'date'-part on
   * the last 'time'; sunset. This means that all dates will have the date of the
   * sunset, thus ensuring equal date across all times of day.
   */
  private readTimes (sunAttributes): TSunCardTimes {
    const sunset = new Date(sunAttributes.next_setting)
    const year = sunset.getUTCFullYear()
    const month = sunset.getUTCMonth()
    const date = sunset.getUTCDate()

    return {
      dawn: this.readTime(sunAttributes.next_dawn, year, month, date),
      dusk: this.readTime(sunAttributes.next_dusk, year, month, date),
      noon: this.readTime(sunAttributes.next_noon, year, month, date),
      sunrise: this.readTime(sunAttributes.next_rising, year, month, date),
      sunset
    }
  }

  private readTime (attributeToParse: string, year: number, month: number, date: number) {
    const read = new Date(attributeToParse)
    read.setUTCFullYear(year)
    read.setUTCMonth(month)
    read.setUTCDate(date)

    return read
  }

  private calculateSunInfo (sunrise: Date, sunset: Date): TSunInfo {
    const sunLine = this.shadowRoot?.querySelector('path') as SVGPathElement

    // find the instances of time for today
    const dayStart = HelperFunctions.todayAtStartOfDay().valueOf()
    const sunriseMs = sunrise.valueOf()
    const sunsetMs = sunset.valueOf()
    const dayEnd = HelperFunctions.todayAtEndOfDay().valueOf()

    // calculate relevant moments in time
    const now = Date.now()
    const msSinceStartOfDay = Math.max(now - dayStart, 0)
    const msSinceDawn = Math.max(now - sunriseMs, 0)
    const msSinceDusk = Math.max(now - sunsetMs, 0)

    const msUntillDawn = sunriseMs - dayStart
    const msOfDaylight = sunsetMs - sunriseMs
    const msUntillEnd = dayEnd - sunsetMs

    // find section positions
    const dawnSectionPosition = HelperFunctions.findSectionPosition(msSinceStartOfDay, msUntillDawn, Constants.SUN_SECTIONS.dawn)
    const daySectionPosition = HelperFunctions.findSectionPosition(msSinceDawn, msOfDaylight, Constants.SUN_SECTIONS.day)
    const duskSectionPosition = HelperFunctions.findSectionPosition(msSinceDusk, msUntillEnd, Constants.SUN_SECTIONS.dusk)

    // find the sun position
    const position = dawnSectionPosition + daySectionPosition + duskSectionPosition
    const sunPosition = sunLine.getPointAtLength(position)

    // calculate section progress, in percentage
    const dawnProgressPercent = HelperFunctions.findSunProgress(
      sunPosition.x, Constants.EVENT_X_POSITIONS.dayStart, Constants.EVENT_X_POSITIONS.sunrise
    )

    const dayProgressPercent = HelperFunctions.findSunProgress(
      sunPosition.x, Constants.EVENT_X_POSITIONS.sunrise, Constants.EVENT_X_POSITIONS.sunset
    )

    const duskProgressPercent = HelperFunctions.findSunProgress(
      sunPosition.x, Constants.EVENT_X_POSITIONS.sunset, Constants.EVENT_X_POSITIONS.dayEnd
    )

    // calculate sun position in regards to the horizon
    const sunCenterY = sunPosition.y - Constants.SUN_RADIUS
    const sunCenterYAboveHorizon = Constants.HORIZON_Y - sunCenterY
    const sunAboveHorizon = sunCenterYAboveHorizon > 0

    let sunPercentOverHorizon = (100 * sunCenterYAboveHorizon) / (2 * Constants.SUN_RADIUS)
    sunPercentOverHorizon = HelperFunctions.clamp(0, 100, sunPercentOverHorizon)

    return {
      sunrise: sunriseMs,
      sunset: sunsetMs,
      dawnProgressPercent,
      dayProgressPercent,
      duskProgressPercent,
      sunAboveHorizon,
      sunPercentOverHorizon,
      sunPosition: { x: sunPosition.x, y: sunPosition.y }
    }
  }
}

window.customCards = window.customCards || []
window.customCards.push({
  type: SunCard.cardType,
  name: SunCard.cardName,
  preview: true,
  description: SunCard.cardDescription
})
