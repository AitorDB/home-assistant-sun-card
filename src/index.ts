import { HomeAssistant } from "custom-card-helpers";
import { customElement, LitElement, state } from "lit-element";

import cardStyles from "./cardStyles";
import { Constants } from "./constants";
import { SunCardContent } from "./cardContent";
import { TSunCardConfig, TSunCardData, TSunCardTime } from "./types";

@customElement("sun-card")
class SunCard extends LitElement {
  static readonly cardType = "sun-card";
  static readonly cardName = "Sun Card";
  static readonly cardDescription =
    "Custom card that display a graph to track the sun position and related events";

  @state()
  private config: TSunCardConfig = {};

  @state()
  private data!: TSunCardData;

  private hasRendered = false;
  private lastHass!: HomeAssistant;

  set hass(hass: HomeAssistant) {
    this.lastHass = hass;

    if (!this.hasRendered) {
      return;
    }

    this.processLastHass();
  }

  calculatePositionAndProgressesByTime(hass: HomeAssistant) {
    const sunLine = this.shadowRoot?.querySelector("path") as SVGPathElement;
    const sunrise = new Date(hass.states["sun.sun"].attributes.next_rising);
    const sunset = new Date(hass.states["sun.sun"].attributes.next_setting);
    const eventsAt = {
      dayStart: 0,
      sunrise: this.convertDateToMinutesSinceDayStarted(sunrise),
      sunset: this.convertDateToMinutesSinceDayStarted(sunset),
      dayEnd: 23 * 60 + 59,
    };

    const now = new Date();
    const minutesSinceTodayStarted =
      this.convertDateToMinutesSinceDayStarted(now);

    // Dawn section position [0 - 105]
    const dawnSectionPosition =
      (Math.min(minutesSinceTodayStarted, eventsAt.sunrise) * 105) /
      eventsAt.sunrise;

    // Day section position [106 - 499]
    const minutesSinceDayStarted = Math.max(
      minutesSinceTodayStarted - eventsAt.sunrise,
      0
    );
    const daySectionPosition =
      (Math.min(minutesSinceDayStarted, eventsAt.sunset - eventsAt.sunrise) *
        (499 - 106)) /
      (eventsAt.sunset - eventsAt.sunrise);

    // Dusk section position [500 - 605]
    const minutesSinceDuskStarted = Math.max(
      minutesSinceTodayStarted - eventsAt.sunset,
      0
    );
    const duskSectionPosition =
      (minutesSinceDuskStarted * (605 - 500)) /
      (eventsAt.dayEnd - eventsAt.sunset);

    const position =
      dawnSectionPosition + daySectionPosition + duskSectionPosition;
    const sunPosition = sunLine.getPointAtLength(position);

    const dawnProgressPercent =
      (100 * (sunPosition.x - Constants.EVENT_X_POSITIONS.dayStart)) /
      (Constants.EVENT_X_POSITIONS.sunrise -
        Constants.EVENT_X_POSITIONS.dayStart);
    const dayProgressPercent =
      (100 * (sunPosition.x - Constants.EVENT_X_POSITIONS.sunrise)) /
      (Constants.EVENT_X_POSITIONS.sunset -
        Constants.EVENT_X_POSITIONS.sunrise);
    const duskProgressPercent =
      (100 * (sunPosition.x - Constants.EVENT_X_POSITIONS.sunset)) /
      (Constants.EVENT_X_POSITIONS.dayEnd - Constants.EVENT_X_POSITIONS.sunset);

    const sunYTop = sunPosition.y - Constants.SUN_RADIUS;
    const yOver = Constants.HORIZON_Y - sunYTop;
    let sunPercentOverHorizon = 0;
    if (yOver > 0) {
      sunPercentOverHorizon = Math.min(
        (100 * yOver) / (2 * Constants.SUN_RADIUS),
        100
      );
    }

    return {
      dawnProgressPercent,
      dayProgressPercent,
      duskProgressPercent,
      sunPercentOverHorizon,
      sunPosition: { x: sunPosition.x, y: sunPosition.y },
    };
  }

  convertDateToMinutesSinceDayStarted(date: Date) {
    return date.getHours() * 60 + date.getMinutes();
  }

  parseTime(timeText: string, locale?: string) {
    const regex = /\d{1,2}[:.]\d{1,2}|[AMP]+/g;
    const date = new Date(timeText);
    const { language, timeFormat } = this.getConfig();
    const result = date
      .toLocaleTimeString(locale ?? language, { hour12: timeFormat === "12h" })
      .match(regex) as [string, ("AM" | "PM")?];

    if (!result && !locale) {
      return this.parseTime(timeText, Constants.DEFAULT_CONFIG.language);
    }

    const [time, period] = result;
    return { time, period };
  }
  calculateTimeBetweenDates(date1: Date, date2: Date): TSunCardTime {
    const timeMs = Math.abs(date1 - date2);
    const { language } = this.getConfig();
    const time = new Date(timeMs).toLocaleTimeString(language, {
      hour12: false,
    });
    return { time };
  }
  processLastHass() {
    if (!this.lastHass) {
      return;
    }

    this.config.darkMode =
      this.config.darkMode ?? this.lastHass.themes.darkMode;
    this.config.language =
      this.config.language ??
      this.lastHass.locale?.language ??
      this.lastHass.language;
    this.config.timeFormat =
      this.config.timeFormat ??
      this.getTimeFormatByLanguage(this.config.language);

    const times = {
      dawn: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_dawn
      ),
      dusk: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_dusk
      ),
      noon: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_noon
      ),
      sunrise: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_rising
      ),
      sunset: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_setting
      ),
    };

    const {
      dawnProgressPercent,
      dayProgressPercent,
      duskProgressPercent,
      sunPercentOverHorizon,
      sunPosition,
    } = this.calculatePositionAndProgressesByTime(this.lastHass);

    const lengthOfDay = this.calculateTimeBetweenDates(
      new Date(this.lastHass.states["sun.sun"].attributes.next_rising),
      new Date(this.lastHass.states["sun.sun"].attributes.next_setting)
    );

    const daylightLeft = this.calculateTimeBetweenDates(
      new Date(this.lastHass.states["sun.sun"].attributes.next_setting),
      new Date()
    );

    const data: TSunCardData = {
      azimuth: this.lastHass.states["sun.sun"].attributes.azimuth,
      dawnProgressPercent,
      dayProgressPercent,
      duskProgressPercent,
      elevation: this.lastHass.states["sun.sun"].attributes.elevation,
      sunPercentOverHorizon,
      sunPosition,
      times,
      lengthOfDay,
      daylightLeft,
    };

    this.data = data;
  }

  getConfig() {
    const config: TSunCardConfig = {};
    config.darkMode = this.config.darkMode ?? Constants.DEFAULT_CONFIG.darkMode;
    config.language = this.config.language ?? Constants.DEFAULT_CONFIG.language;
    config.showAzimuth =
      this.config.showAzimuth ?? Constants.DEFAULT_CONFIG.showAzimuth;
    config.showElevation =
      this.config.showElevation ?? Constants.DEFAULT_CONFIG.showElevation;
    config.timeFormat =
      this.config.timeFormat ?? Constants.DEFAULT_CONFIG.timeFormat;
    config.title = this.config.title;

    if (
      !Object.keys(Constants.LOCALIZATION_LANGUAGES).includes(config.language!)
    ) {
      config.language = Constants.DEFAULT_CONFIG.language;
    }

    return config;
  }

  getTimeFormatByLanguage(language: string) {
    const date = new Date();
    const time = date.toLocaleTimeString(language).toLocaleLowerCase();
    return time.includes("pm") || time.includes("am") ? "12h" : "24h";
  }

  setConfig(config: TSunCardConfig) {
    this.config = { ...config };
  }

  protected render() {
    const config = this.getConfig();
    const language = config.language!;
    const localization = Constants.LOCALIZATION_LANGUAGES[language];
    return SunCardContent.generate(this.data, localization, config);
  }

  protected updated(changedProperties) {
    super.updated(changedProperties);

    if (!this.hasRendered) {
      this.hasRendered = true;
      this.processLastHass();
    }
  }

  static get styles() {
    return cardStyles;
  }
}

window.customCards = window.customCards || [];
window.customCards.push({
  type: SunCard.cardType,
  name: SunCard.cardName,
  description: SunCard.cardDescription,
});
