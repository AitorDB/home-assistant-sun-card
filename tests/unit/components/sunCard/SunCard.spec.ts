import { HomeAssistant } from 'custom-card-helpers'
import { css, CSSResult } from 'lit-element'
import { SunCard } from '../../../../src/components/sunCard/SunCard'
import { SunCardEditor } from '../../../../src/components/sunCardEditor'
import { Constants } from '../../../../src/constants'
import { ESunCardErrors, ISunCardConfig, TSunCardData } from '../../../../src/types'
import { TemplateResultTestHelper } from '../../../helpers/TestHelpers'

jest.mock('../../../../src/components/SunErrorContent', () => require('../../../mocks/SunErrorContent'))
jest.mock('../../../../src/components/sunCard/SunCardContent', () => require('../../../mocks/SunCardContent'))
jest.mock('../../../../src/utils/HelperFunctions', () => require('../../../mocks/HelperFunctions'))
jest.mock('../../../../src/utils/I18N', () => require('../../../mocks/I18N'))
jest.mock('../../../../src/cardStyles', () => css``)


// JSDom doesn't include SVGPathElement
class SVGPathElement {
  getPointAtLength () {
    return {
      x: 0,
      y: 0
    }
  }
}

Object.defineProperty(window, 'SVGPathElement' , { value: SVGPathElement })

describe('SunCard', () => {
  let sunCard: SunCard

  beforeEach(() => {
    sunCard = new SunCard()
  })

  describe('set hass', () => {
    it('updates lastHass property', () => {
      jest.spyOn(sunCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      jest.spyOn(sunCard as any, 'processLastHass').mockReturnValue(undefined)
      expect(sunCard['lastHass']).toBeUndefined()

      const expectedLastHass = {} as HomeAssistant
      sunCard.hass = expectedLastHass
      expect(sunCard['lastHass']).toEqual(expectedLastHass)
    })

    it('calls populateConfigFromHass if it has not been rendered yet', () => {
      const populateConfigFromHassSpy = jest.spyOn(sunCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      jest.spyOn(sunCard as any, 'processLastHass').mockReturnValue(undefined)
      sunCard['hasRendered'] = false

      sunCard.hass = {} as HomeAssistant
      expect(populateConfigFromHassSpy).toHaveBeenCalledTimes(1)
    })

    it('does not call populateConfigFromHass if it has been rendered', () => {
      const populateConfigFromHassSpy = jest.spyOn(sunCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      jest.spyOn(sunCard as any, 'processLastHass').mockReturnValue(undefined)
      sunCard['hasRendered'] = true

      sunCard.hass = {} as HomeAssistant
      expect(populateConfigFromHassSpy).not.toHaveBeenCalled()
    })

    it('calls processLastHass if it has been rendered', () => {
      jest.spyOn(sunCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      const processLastHassSpy = jest.spyOn(sunCard as any, 'processLastHass').mockReturnValue(undefined)
      sunCard['hasRendered'] = true

      sunCard.hass = {} as HomeAssistant
      expect(processLastHassSpy).toHaveBeenCalledTimes(1)
    })
  })

  describe('getConfigElement', () => {
    it('creates and return a sun card config element', () => {
      const expectedElement = document.createElement('div')
      const createElementSpy = jest.spyOn(document, 'createElement').mockReturnValueOnce(expectedElement)
      const result = SunCard.getConfigElement()

      expect(result).toEqual(expectedElement)
      expect(createElementSpy).toHaveBeenCalledTimes(1)
      expect(createElementSpy).toHaveBeenCalledWith(SunCardEditor.cardType)
    })
  })

  describe('setConfig', () => {
    it('overrides old config with new values', () => {
      jest.spyOn(sunCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      expect(sunCard['config']).toEqual({ type: SunCard.cardType })
      const config = {
        type: SunCard.cardType,
        title: 'Test'
      } as ISunCardConfig

      sunCard.setConfig(config)
      expect(sunCard['config'].title).toEqual(config.title)

      const newConfig = {
        type: SunCard.cardType,
        title: 'Test2'
      } as ISunCardConfig

      sunCard.setConfig(newConfig)
      expect(sunCard['config'].title).toEqual(newConfig.title)
    })

    it('uses the provided component', () => {
      const config = {
        type: SunCard.cardType,
        component: 'test'
      } as ISunCardConfig

      sunCard.setConfig(config)
      expect(sunCard['config'].component).toEqual(config.component)
    })

    it('uses the default component when not provided', () => {
      const config = {
        type: SunCard.cardType
      } as ISunCardConfig

      sunCard.setConfig(config)
      expect(sunCard['config'].component).toEqual(Constants.DEFAULT_CONFIG.component)
    })

    it('throws an error if the provided language is not supported', () => {
      const config = {
        type: SunCard.cardType,
        language: 'notSupportedLanguage'
      } as ISunCardConfig

      let thrownError
      try {
        sunCard.setConfig(config)
      } catch (error) {
        thrownError = error.message
      }

      expect(thrownError).toEqual(`${config.language} is not a supported language. Supported languages: ${Object.keys(Constants.LOCALIZATION_LANGUAGES)}`)
    })

    const fields = ['sunrise', 'sunset', 'dawn', 'noon', 'dusk', 'azimuth', 'elevation']
    for (const field of fields) {
      it(`uses the provided value for ${field}`, () => {
        const config = {
          type: SunCard.cardType,
          fields: {
            [field]: 'test'
          }
        } as ISunCardConfig

        sunCard.setConfig(config)
        expect(sunCard['config'].fields?.[field]).toEqual('test')
      })

      it(`uses the default value for ${field} when not provided`, () => {
        const config = {
          type: SunCard.cardType
        } as ISunCardConfig

        sunCard.setConfig(config)
        expect(sunCard['config'].fields?.[field]).toEqual(Constants.DEFAULT_CONFIG.fields?.[field])
      })
    }

    it('calls populateConfigFromHass if lastHass has a value', () => {
      const populateConfigFromHassSpy = jest.spyOn(sunCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      sunCard['lastHass'] = {} as HomeAssistant
      
      const config = {
        type: SunCard.cardType
      } as ISunCardConfig

      sunCard.setConfig(config)
      expect(populateConfigFromHassSpy).toHaveBeenCalledTimes(1)
    })

    it('does not call populateConfigFromHass if lastHass has not a value', () => {
      const populateConfigFromHassSpy = jest.spyOn(sunCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      delete (sunCard as any)['lastHass']
      
      const config = {
        type: SunCard.cardType
      } as ISunCardConfig

      sunCard.setConfig(config)
      expect(populateConfigFromHassSpy).not.toHaveBeenCalled()
    })
  })

  describe('render', () => {
    it('renders an error if error is present on data', async () => {
      if (!sunCard['data']) {
        sunCard['data'] = {} as TSunCardData
      }

      sunCard['data'].error = ESunCardErrors.SunIntegrationNotFound
      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof sunCard.render>
      element.templateResultFunction = () => sunCard.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })

    it('renders the sun card if no error is present on data', async () => {
      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof sunCard.render>
      element.templateResultFunction = () => sunCard.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })
  })

  describe('updated', () => {
    it('sets to true hasRendered if has not been rendered yet', () => {
      jest.spyOn(sunCard as any, 'processLastHass').mockReturnValue(undefined)
      sunCard['hasRendered'] = false
      sunCard['updated'](new Map())
      expect(sunCard['hasRendered']).toEqual(true)
    })

    it('calls processLastHass if has not been rendered yet', () => {
      const processLastHassSpy = jest.spyOn(sunCard as any, 'processLastHass').mockReturnValue(undefined)
      sunCard['hasRendered'] = false
      sunCard['updated'](new Map())
      expect(processLastHassSpy).toHaveBeenCalledTimes(1)
    })

    it('does nothing more than call super.updated if has been already rendered', () => {
      const processLastHassSpy = jest.spyOn(sunCard as any, 'processLastHass').mockReturnValue(undefined)
      sunCard['hasRendered'] = true
      sunCard['updated'](new Map())
      expect(processLastHassSpy).not.toHaveBeenCalledWith()
    })
  })

  describe('populateConfigFromHass', () => {
    it('keeps values from config when present', () => {
      const config = {
        type: SunCard.cardType,
        darkMode: true,
        language: 'es'
      } as ISunCardConfig

      sunCard['lastHass'] = {
        themes: {
          darkMode: false
        },
        locale: {
          language: 'en'
        }
      } as any

      sunCard['config'] = config
      sunCard['populateConfigFromHass']()
      expect(sunCard['config'].darkMode).toEqual(config.darkMode)
      expect(sunCard['config'].language).toEqual(config.language)
    })

    it('sets values from hass when not present in config', () => {
      sunCard['lastHass'] = {
        themes: {
          darkMode: false
        },
        locale: {
          language: 'it'
        }
      } as any

      sunCard['populateConfigFromHass']()
      expect(sunCard['config'].darkMode).toEqual(false)
      expect(sunCard['config'].language).toEqual('it')
    })

    it('supports old versions of home assistant', () => {
      sunCard['lastHass'] = {
        themes: {
          darkMode: false
        },
        language: 'it'
      } as any

      sunCard['populateConfigFromHass']()
      expect(sunCard['config'].darkMode).toEqual(false)
      expect(sunCard['config'].language).toEqual('it')
    })
  })

  describe('processLastHass', () => {
    it('does an early return if lastHass has no value', () => {
      delete (sunCard as any)['lastHass']
      const populateConfigFromHassSpy = jest.spyOn(sunCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      sunCard['processLastHass']()

      expect(populateConfigFromHassSpy).not.toHaveBeenCalled()
    })

    it('calls populateConfigFromHass', () => {
      (sunCard as any)['lastHass'] = { states: {} }
      const populateConfigFromHassSpy = jest.spyOn(sunCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      sunCard['processLastHass']()

      expect(populateConfigFromHassSpy).toHaveBeenCalledTimes(1)
    })

    it('process the data from the component when found', () => {
      (sunCard as any)['lastHass'] = {
        states: {
          component: {
            attributes: {
              azimuth: 3,
              elevation: 7
            }
          }
        }
      }

      const config = {
        type: SunCard.cardType,
        component: 'component'
      } as ISunCardConfig

      sunCard['config'] = config

      jest.spyOn(sunCard as any, 'populateConfigFromHass').mockReturnValue(undefined)

      const times = {
        dawn: new Date(0),
        dusk: new Date(0),
        noon: new Date(0),
        sunrise: new Date(0),
        sunset: new Date(0)
      }

      const sunInfo = {
        sunrise: 0,
        sunset: 0,
        dawnProgressPercent: 0,
        dayProgressPercent: 0,
        duskProgressPercent: 0,
        sunAboveHorizon: 0,
        sunPercentOverHorizon: 0,
        sunPosition: { x: 0, y: 0 }
      }

      const readTimesSpy = jest.spyOn(sunCard as any, 'readTimes').mockReturnValue(times)
      const calculateSunInfoSpy = jest.spyOn(sunCard as any, 'calculateSunInfo').mockReturnValue(sunInfo)

      sunCard['processLastHass']()

      expect(readTimesSpy).toHaveBeenCalledTimes(1)
      expect(calculateSunInfoSpy).toHaveBeenCalledTimes(1)
      expect(sunCard['data'].times).toEqual(times)
      expect(sunCard['data'].sunInfo).toEqual(sunInfo)
    })

    it('sets an error if the component is not found', () => {
      (sunCard as any)['lastHass'] = { states: {} }
      jest.spyOn(sunCard as any, 'populateConfigFromHass').mockReturnValue(undefined)
      sunCard['processLastHass']()

      expect(sunCard['data'].error).toEqual(ESunCardErrors.SunIntegrationNotFound)
    })
  })

  describe('readTimes', () => {
    it('returns dates for each time field', () => {
      const readTimeSpy = jest.spyOn(sunCard as any, 'readTime').mockReturnValue(new Date(0))
      const result = sunCard['readTimes']({
        next_setting: 0,
        next_dawn: 0,
        next_dusk: 0,
        next_noon: 0,
        next_rising: 0
      })
  
      expect(result).toEqual({
        dawn: new Date(0),
        dusk: new Date(0),
        noon: new Date(0),
        sunrise: new Date(0),
        sunset: new Date(0)
      })
  
      expect(readTimeSpy).toHaveBeenCalledTimes(4)
    })
  })

  describe('readTime', () => {
    it('sets a specific day, month and year to a provided string date', () => {
      const result = sunCard['readTime']('0', 2021, 5, 12)
      expect(result).toEqual(new Date(1623456000000)) // Sat Jun 12 2021 01:00:00 GMT+0100 (British Summer Time)
    })
  })

  describe('calculateSunInfo', () => {
    it('returns all sun info', () => {
      const path = new SVGPathElement()
      jest.spyOn((sunCard as any).shadowRoot, 'querySelector').mockReturnValue(path)

      const result = sunCard['calculateSunInfo'](new Date(0), new Date(0))
      expect(result).toEqual({
        dawnProgressPercent: 0,
        dayProgressPercent: 0,
        duskProgressPercent: 0,
        sunAboveHorizon: true,
        sunPercentOverHorizon: 0,
        sunPosition: {
         x: 0,
         y: 0
        },
        sunrise: 0,
        sunset: 0
      })
    })
  })

  describe('get styles', () => {
    it('returns a CSSResult', () => {
      expect(SunCard.styles).toBeInstanceOf(CSSResult)
    })
  })
})
