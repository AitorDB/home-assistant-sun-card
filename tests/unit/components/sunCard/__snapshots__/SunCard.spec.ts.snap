// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SunCard render renders an error if error is present on data 1`] = `
<!---->
      <div class="sun-card-error">
        ERROR
      </div>
    
`;

exports[`SunCard render renders the sun card if no error is present on data 1`] = `
<!---->
      <ha-card>
        <div class="sun-card">
          SUN CARD CONTENT
        </div>
      </ha-card>
    
`;
